openedx_learning/apps/authoring/containers/api.py

```diff
@@ -0,0 +1,451 @@
+            created_by=created_by,
+        )
+    return (container, container_version)
+
+
+def get_container(pk: int) -> Container:
+    """
+    [ 🛑 UNSTABLE ]
+    Get a container by its primary key.
+
+    Args:
+        pk: The primary key of the container.
+
+    Returns:
+        The container with the given primary key.
+    """
+    # TODO: should this use with_publishing_relations as in components?
+    return Container.objects.get(pk=pk)
+
+
+@dataclass(frozen=True)
+class ContainerEntityListEntry:
+    """
+    [ 🛑 UNSTABLE ]
+    Data about a single entity in a container, e.g. a component in a unit.
+    """
+    entity_version: PublishableEntityVersion
+    pinned: bool
+
+    @property
+    def entity(self):
+        return self.entity_version.entity
+
+
+def get_entities_in_draft_container(
+    container: Container | ContainerMixin,
+) -> list[ContainerEntityListEntry]:
+    """
+    [ 🛑 UNSTABLE ]
+    Get the list of entities and their versions in the draft version of the
+    given container.
+    """
+    if isinstance(container, ContainerMixin):
+        container = container.container
+    assert isinstance(container, Container)
+    entity_list = []
+    for row in container.versioning.draft.entity_list.entitylistrow_set.order_by("order_num"):
+        entity_version = row.entity_version or row.entity.draft.version
+        if entity_version is not None:  # As long as this hasn't been soft-deleted:
+            entity_list.append(ContainerEntityListEntry(
+                entity_version=row.entity_version or row.entity.draft.version,
+                pinned=row.entity_version is not None,
+            ))
+        # else should we indicate somehow a deleted item was here?
+    return entity_list
+
+
+def get_entities_in_published_container(
+    container: Container | ContainerMixin,
+) -> list[ContainerEntityListEntry] | None:
+    """
+    [ 🛑 UNSTABLE ]
+    Get the list of entities and their versions in the draft version of the
+    given container.
+    """
+    if isinstance(container, ContainerMixin):
+        cv = container.container.versioning.published
+    elif isinstance(container, Container):
+        cv = container.versioning.published
+    else:
+        raise TypeError(f"Expected Container or ContainerMixin; got {type(container)}")
+    if cv is None:
+        return None  # There is no published version of this container. Should this be an exception?
+    assert isinstance(cv, ContainerVersion)
+    entity_list = []
+    for row in cv.entity_list.entitylistrow_set.order_by("order_num"):
+        entity_version = row.entity_version or row.entity.published.version
+        if entity_version is not None:  # As long as this hasn't been soft-deleted:
+            entity_list.append(ContainerEntityListEntry(
+                entity_version=entity_version,
+                pinned=row.entity_version is not None,
+            ))
+        # else should we indicate somehow a deleted item was here?
```

Comment:
Maybe `ContainerEntityListEntry` could handle when the entry was previously soft-deleted?

Reply:
Yes that's what I was thinking, but we don't have any use case for it yet.